File: src/cli/args.js
// Handwritten argv parser. Returns { positional, flags } where flags is a plain
// object (booleans for `--name`/`--no-name`, strings for `--name value` or
// `--name=value`, comma-split arrays for repeatable flags via repeat()). Stops
// parsing flags after `--`. Reserved globals (`--json`, `--help`, `--version`)
// are not consumed here — the entry point reads them before delegating.

const REPEATABLE = new Set(["tag", "tags"]);

function isFlag(token) {
  return token.startsWith("--") && token.length > 2;
}

function splitFlag(token) {
  const eq = token.indexOf("=");
  if (eq === -1) return { name: token.slice(2), value: undefined };
  return { name: token.slice(2, eq), value: token.slice(eq + 1) };
}

function camelize(name) {
  return name.replace(/-([a-z])/g, (_, ch) => ch.toUpperCase());
}

function coerceScalar(value) {
  if (value === "true") return true;
  if (value === "false") return false;
  if (/^-?\d+$/.test(value)) return Number(value);
  if (/^-?\d+\.\d+$/.test(value)) return Number(value);
  return value;
}

function assignFlag(flags, name, rawValue) {
  if (name.startsWith("no-") && rawValue === undefined) {
    flags[camelize(name.slice(3))] = false;
    return;
  }
  const value = rawValue === undefined ? true : coerceScalar(rawValue);
  if (REPEATABLE.has(name) && value !== true && value !== false) {
    const list = String(value)
      .split(",")
      .map((v) => v.trim())
      .filter(Boolean);
    const key = camelize(name);
    flags[key] = Array.isArray(flags[key]) ? [...flags[key], ...list] : list;
    return;
  }
  flags[camelize(name)] = value;
}

export function parseArgs(argv) {
  const positional = [];
  const flags = {};
  let i = 0;
  let stopFlags = false;
  while (i < argv.length) {
    const token = argv[i];
    if (stopFlags) {
      positional.push(token);
      i += 1;
      continue;
    }
    if (token === "--") {
      stopFlags = true;
      i += 1;
      continue;
    }
    if (isFlag(token)) {
      const { name, value } = splitFlag(token);
      if (value !== undefined) {
        assignFlag(flags, name, value);
        i += 1;
        continue;
      }
      const next = argv[i + 1];
      if (next !== undefined && !isFlag(next)) {
        assignFlag(flags, name, next);
        i += 2;
        continue;
      }
      assignFlag(flags, name, undefined);
      i += 1;
      continue;
    }
    positional.push(token);
    i += 1;
  }
  return { positional, flags };
}

File: atlas-py/atlas_py/cli/args.py
"""Handwritten argv parser. Returns {positional, flags} where flags is a plain
object (booleans for --name/--no-name, strings for --name value or
--name=value, comma-split arrays for repeatable flags). Stops parsing flags
after --. Mirrors src/cli/args.js.
"""

from __future__ import annotations

import re

REPEATABLE = {"tag", "tags"}

_INT_RE = re.compile(r"^[+-]?\d+$")
_FLOAT_RE = re.compile(r"^[+-]?\d+\.\d+$")


def _is_flag(token):
    return token.startswith("--") and len(token) > 2


def _split_flag(token):
    eq = token.find("=")
    if eq == -1:
        return token[2:], None
    return token[2:eq], token[eq + 1:]


def _camelize(name):
    parts = name.split("-")
    return parts[0] + "".join(p[:1].upper() + p[1:] for p in parts[1:] if p)


def _coerce_scalar(value):
    if value == "true":
        return True
    if value == "false":
        return False
    if _INT_RE.match(value):
        return int(value)
    if _FLOAT_RE.match(value):
        return float(value)
    return value


def _assign_flag(flags, name, raw_value):
    if name.startswith("no-") and raw_value is None:
        flags[_camelize(name[3:])] = False
        return
    value = True if raw_value is None else _coerce_scalar(raw_value)
    if name in REPEATABLE and value is not True and value is not False:
        items = [v.strip() for v in str(value).split(",") if v.strip()]
        key = _camelize(name)
        existing = flags.get(key)
        if isinstance(existing, list):
            flags[key] = existing + items
        else:
            flags[key] = items
        return
    flags[_camelize(name)] = value


def parse_args(argv):
    """Parse argv tokens into positional + flags dict."""
    positional = []
    flags = {}
    i = 0
    stop_flags = False
    while i < len(argv):
        token = argv[i]
        if stop_flags:
            positional.append(token)
            i += 1
            continue
        if token == "--":
            stop_flags = True
            i += 1
            continue
        if _is_flag(token):
            name, value = _split_flag(token)
            if value is not None:
                _assign_flag(flags, name, value)
                i += 1
                continue
            nxt = argv[i + 1] if i + 1 < len(argv) else None
            if nxt is not None and not _is_flag(nxt):
                _assign_flag(flags, name, nxt)
                i += 2
                continue
            _assign_flag(flags, name, None)
            i += 1
            continue
        positional.append(token)
        i += 1
    return {"positional": positional, "flags": flags}
